[Yan Xiaoying] iP

xiaoying1129 opened the pull request:
# Duke 🤖 
> Life is like riding a bicycle. To keep your balance, you must keep moving. - Albert Einstein ([source](https://parade.com/1240718/kelseypelzer/albert-einstein-quotes/))

## Commands
Duke is a Personal Assistant Chatbot that helps you keep track of your tasks and deadlines. As of now, Duke supports the following commands:
- **List:** list all the tasks stored in the list
- **Mark:** mark a task as done ✔️ 
- **Unmark:** mark a task as not done yet
- **Todo:** add tasks without any date/time attached to it, e.g. `todo borrow book`
- **Deadline:** add tasks that need to be done before a specific date/time, e.g. `deadline return book /by Sunday`
- **Event:** add tasks that start at a specific time and ends at a specific time, e.g. `event project meeting /at Mon 2-4pm`
- **Delete:** delete a task
- **Find:** find a task by searching for a keyword, e.g. `find book`

## Start using Duke
1. To run Duke, simply download the *jar* file from **releases**.
2. Launch it by double-clicking the file or run the following commands in your terminal.
```
cd [file path to the folder containing the jar file]
java -jar duke.jar
```
3. Add your tasks. 📁 
4. Let it manage your tasks for you! 🎉 

## GFMD Elements
This update contains the following required GFMD elements:
- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list
- [x] an emoji 😉 
- [x] a blockquote
- [x] a hyper link
- [x] inline code
- [x] some text formatting: bold, italic, stikethrough etc.

Commit:
Add Gradle support

Commit:
Implemented Level-1

Commit:
Implemented Level-2

Commit:
Implemented Level-3

Commit:
Implemented Level-4

Commit:
Automated Text UI Testing

Commit:
Implemented Level-5

Commit:
Implemented Level-6

Commit:
implemented Level-7

Commit:
Implemented level-8

xiaoying1129 committed:
Merge pull request #1 from xiaoying1129/branch-Level-7

implemented Level-7

Commit:
Merge branch 'master' into branch-Level-8

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java

Commit:
Merge branch 'master' of https://github.com/xiaoying1129/ip into branch-Level-8

xiaoying1129 committed:
Merge pull request #2 from xiaoying1129/branch-Level-8

Branch level 8

Commit:
fixed task access

Commit:
Implemented A-MoreOOP

Commit:
Implemented A-Packages

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'

Commit:
Add JUnit Tests

Commit:
Package the App as a JAR file

Commit:
Add JavaDoc comments

Commit:
Tweak the code to comply with a coding standard

Commit:
Give users a way to find a task by searching for a keyword.

Commit:
Merge branch 'master' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/duke/command/DeleteCommand.java

Commit:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/duke/main/Parser.java

xiaoying1129 committed:
Merge pull request #3 from xiaoying1129/branch-A-CodingStandard

Tweak the code to comply with a coding standard

xiaoying1129 committed:
Merge pull request #4 from xiaoying1129/branch-Level-9

Give users a way to find a task by searching for a keyword.

sheshenk commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            return new ListCommand();
+        } else if (input.startsWith("mark ")) {
+            int taskNum = Integer.parseInt(input.replace("mark ", ""));
+            validateMark(taskNum);
```
I like that you used validation functions instead of validating within the `parse` method.

sheshenk commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     * @throws DukeException  If illegal input is entered.
+     */
+    static Command parse(String input) throws DukeException {
+        if (input.equals("bye")) {
+            return new ExitCommand();
+        } else if (input.equals("list")) {
+            return new ListCommand();
+        } else if (input.startsWith("mark ")) {
+            int taskNum = Integer.parseInt(input.replace("mark ", ""));
+            validateMark(taskNum);
+            return new DoneCommand(taskNum);
```
Perhaps this could be replaced with a `switch` statement? The direction from the input (`unmark`, `deadline`, etc.) is usually one word, so you can use `split` with the `limit` argument.

sheshenk commented on src/main/java/duke/main/Parser.java:
```diff
@@ -0,0 +1,134 @@
+        if(toFind.isEmpty()) {
+            throw new DukeException("OOPS!!! The task to find cannot be empty.");
+        }
+    }
```
Maybe a variable name like `query` could be used here?

sheshenk commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,101 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
I like the way the imports are grouped by package!

sheshenk commented on src/main/java/duke/main/Storage.java:
```diff
@@ -0,0 +1,101 @@
+     * @param tasks list of tasks to be saved
+     * @throws DukeException  If there is an error.
+     */
+    public void save(ArrayList<Task> tasks) throws DukeException {
+        try {
+            File file = new File(filePath);
+
+            FileWriter fw = new FileWriter(filePath);
+            for (Task task : tasks) {
+                fw.write(task.getOutput() + "\n");
+            }
+            fw.close();
+        } catch (IOException e) {
+            throw new DukeException("An IOException occurred!! :(");
+        }
+    }
+}
```
Perhaps if the save file is not found, it could be created?

sheshenk commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Prints a line divider.
+     */
+    public void linePrint() {
+        System.out.println("\t____________________________________________________________");
+    }
```
You could use static variables to store the line, indent, line break constants.

sheshenk commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String toString() {
+        if (byDate != null) {
+            return "[D]" + super.toString() + " (by: " + byDate.format(DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm")) + ")";
+        } else {
+            return "[D]" + super.toString() + " (by: " + by + ")";
+        }
```
Perhaps `String.format` could be using to simplify the concatenation?

sheshenk commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,40 @@
+
+public class Task {
+    String description;
+    boolean isDone;
```
I like the way your booleans are all consistent and sound like booleans!

sheshenk commented on src/test/java/duke/main/ParserTest.java:
```diff
@@ -0,0 +1,23 @@
+package duke.main;
+
+import duke.command.*;
```
Maybe the wildcard import can be avoided?

sheshenk reviewed (commented):
Overall, this is a really well-written PR. I like how all the commands have been abstracted into their own classes.

yilinzyl commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,19 @@
+
+public class Event extends Task {
+    protected String time;
+
```
Maybe this could be named at to follow the same naming format used in the Deadline class?

yilinzyl commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Prints a greeting message when the bot is opened.
+     */
+    public void greetingMessage() {
```
Maybe this method could be named like a verb, eg. showGreetingMessage() instead?

yilinzyl commented on src/main/java/duke/main/Ui.java:
```diff
@@ -0,0 +1,92 @@
+    /**
+     * Prints an exit message.
+     */
+    public void exitMessage() {
```
Maybe this method name could also be renamed as showExitMessage() to sound more like a verb?

yilinzyl commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,19 @@
+    public Event(String description, String time) {
+        super(description);
+        this.time = time;
+    }
```
Any reason why you chose to have 2 constructors (string and datetime) for Deadline but not Event?

yilinzyl commented on src/main/java/duke/main/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+ * Contains the task list and relevant operations on tasks.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
```
I like that your collection, tasks, is named in plural form!

yilinzyl reviewed (commented):
Overall clean code! Just a few naming conventions to consider :)

Commit:
fixed coding style violations

xiaoying1129 committed:
Merge pull request #5 from xiaoying1129/branch-Level-9

fixed coding style violations

Commit:
Added a GUI to Duke.

xiaoying1129 committed:
Merge branch 'branch-Level-10'

Added a GUI to Duke.

Commit:
Add assertions to ensure user input is validated.

xiaoying1129 committed:
Add assertions to ensure user input is validated.

Assertions are added to ensure that user input is validated by the validation functions.

Unvalidated user input can lead to bugs.

Assertions are useful in documenting important assumptions that should hold at various points in the co

Commit:
Improve code quality where necessary.

xiaoying1129 committed:
Merge pull request #8 from xiaoying1129/branch-A-CodeQuality

Improve code quality where necessary.

Commit:
Add extension C-Tagging

xiaoying1129 committed:
Merge pull request #10 from xiaoying1129/branch-C-Tagging

Add extension C-Tagging

Commit:
Improve GUI

Commit:
Tweak phrases used by Duke

Commit:
Add user guide and Ui screenshot

Commit:
Fix build.gradle
